File: 3c.cpp
#include<iostream>

using namespace std;

char a[3][3];

bool win(char c){
	for(int i=0;i<3;i++){
		if(a[i][0] == c && a[i][1] == c && a[i][2] == c)
			return true;
		if(a[0][i] == c && a[1][i] == c && a[2][i] == c)
			return true;
	}
	
	if(a[0][0] == c && a[1][1] == c && a[2][2] == c)
		return true;
	if(a[0][2] == c && a[1][1] == c && a[2][0] == c)
		return true;
	
	return false;
}


int main(){
	int xct = 0, oct=0;
 
 
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++) {
        	cin>>a[i][j];
            xct += a[i][j] == 'X';
            oct += a[i][j] == '0';
        }
 
    if (xct<oct or xct-oct>1 or (win('X')&&xct-oct!=1) or (win('0') && xct!=oct)) {
        printf("illegal");
    }
 
    else if (win('X'))
        printf("the first player won");
 
    else if (win('0'))
        printf("the second player won");
 
    else if (xct + oct == 9)
        printf("draw");
 
    else if (xct == oct)
        printf("first");
 
    else
        printf("second");
       
 
    return 0;
}

File: 3c.py
def win(l,s):
    if l[0] == [s, s, s] or l[1] == [s, s, s] or l[2] == [s, s, s] or l[0][0] == s and l[1][0] == s and l[2][0] == s or \
            l[0][1] == s and l[1][1] == s and l[2][1] == s or l[0][2] == s and l[1][2] == s and l[2][2] == s or l[0][
        0] == s and l[1][1] == s and l[2][2] == s or l[0][2] == s and l[1][1] == s and l[2][0] == s:
        return 1
    else:
        return 0
a=list(input())
b=list(input())
c=list(input())
l=[a,b,c]
mf=a.count('X')+b.count('X')+c.count('X')
ms=a.count('0')+b.count('0')+c.count('0')
me=a.count('.')+b.count('.')+c.count('.')
if  mf -ms!= 1 and mf -ms!=0 :
    print('illegal')
else:
    n=win(l,'0')
    m=win(l,'X')
    if n==1 and m==1 or n==1 and mf-ms!=0 or m==1 and mf-ms!=1:
        print('illegal')
    elif n==1:
        print('the second player won')
    elif m==1:
        print('the first player won')
    else:
        if me==0:
            print('draw')
        elif mf-ms==1:
            print('second')
        else:print('first')
